File: api/src/main/java/com/castlefrog/agl/domains/havannah/HavannahAction.java
package com.castlefrog.agl.domains.havannah;

import java.io.Serializable;
import java.util.Vector;

import com.castlefrog.agl.Action;

public final class HavannahAction implements Action, Serializable {
    /** list of all possible Havannah actions */
    private static final Vector<Vector<HavannahAction>> actions_ = new Vector<>();

    /** x coordinate */
    private final byte x_;
    /** y coordinate */
    private final byte y_;

    private HavannahAction(int x, int y) {
        x_ = (byte) x;
        y_ = (byte) y;
    }

    public static HavannahAction valueOf(int x, int y) {
        if (x >= actions_.size() || y >= actions_.size()) {
            generateActions(Math.max(x + 1, y + 1));
        }
        return actions_.get(x).get(y);
    }

    public HavannahAction copy() {
        return this;
    }

    /**
     * Generates all possible actions and a few impossible actions.
     */
    private static void generateActions(int size) {
        actions_.setSize(size);
        for (int i = 0; i < size; i += 1) {
            if (actions_.get(i) == null) {
                actions_.set(i, new Vector<HavannahAction>());
            }
            actions_.get(i).setSize(size);
            for (int j = 0; j < size; j += 1) {
                if (actions_.get(i).get(j) == null) {
                    actions_.get(i).set(j, new HavannahAction(i, j));
                }
            }
        }
    }

    public int getX() {
        return x_;
    }

    public int getY() {
        return y_;
    }

    @Override
    public int hashCode() {
        return 11 * (7 + x_) + y_;
    }

    @Override
    public boolean equals(Object object) {
        if (!(object instanceof HavannahAction)) {
            return false;
        }
        HavannahAction action = (HavannahAction) object;
        return x_ == action.getX() && y_ == action.getY();
    }

    @Override
    public String toString() {
        return "" + ((char) (0x41 + x_)) + y_;
    }
}

File: api/src/main/java/com/castlefrog/agl/domains/havannah/HavannahAction.kt
package com.castlefrog.agl.domains.havannah

import java.io.Serializable
import java.util.Vector

import com.castlefrog.agl.Action

data class HavannahAction private constructor(val x: Byte, val y: Byte) : Action, Serializable {

    override fun copy(): HavannahAction {
        return this
    }

    override fun toString(): String {
        return "" + (0x41 + x).toChar() + y
    }

    companion object {
        private val actions = Vector<Vector<HavannahAction>>()

        /**
         * Returns a havannah action.
         * @param x x-coordinate
         * @param y y-coordinate
         * @return action corresponding to x and y coordinate
         */
        fun valueOf(x: Int, y: Int): HavannahAction {
            if (x >= actions.size || y >= actions.size) {
                generateActions(Math.max(x + 1, y + 1))
            }
            return actions[x][y]
        }

        private fun generateActions(size: Int) {
            actions.setSize(size)
            for (i in 0..size - 1) {
                if (actions[i] == null) {
                    actions[i] = Vector<HavannahAction>()
                }
                actions[i].setSize(size)
                for (j in 0..size - 1) {
                    if (actions[i][j] == null) {
                        actions[i][j] = HavannahAction(i.toByte(), j.toByte())
                    }
                }
            }
        }
    }
}
